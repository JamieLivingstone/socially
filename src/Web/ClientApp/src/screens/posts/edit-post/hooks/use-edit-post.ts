import { useToast } from '@chakra-ui/react';
import axios from 'axios';
import { useMutation, useQueryClient } from 'react-query';
import { useNavigate } from 'react-router-dom';

import { useAuth } from '@hooks/use-auth';

export function useEditPost() {
  const { account } = useAuth();
  const queryClient = useQueryClient();
  const navigate = useNavigate();
  const toast = useToast();

  const { mutateAsync } = useMutation(
    async function createPostRequest(payload: { title: string; body: string; slug: string; tags: Array<string> }) {
      try {
        await axios.patch(`/api/v1/posts/${payload.slug}`, payload);
        await queryClient.invalidateQueries(['post', payload.slug]);
        navigate(`/${account?.username}/${payload.slug}`);
      } catch (error) {
        toast({
          title: 'Failed to edit post!',
          status: 'error',
          duration: 3000,
          isClosable: false,
        });
      }
    },
    { useErrorBoundary: false },
  );

  return {
    editPost: mutateAsync,
  };
}
