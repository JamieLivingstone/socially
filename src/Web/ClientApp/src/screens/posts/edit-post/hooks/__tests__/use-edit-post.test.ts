import { renderHook } from '@testing-library/react-hooks';
import moxios from 'moxios';

import Providers from '@providers';

import { useEditPost } from '../use-edit-post';

describe('useEditPost', () => {
  test('edits post and navigates to slug', async () => {
    moxios.stubRequest('/api/v1/posts/test-post', {
      status: 200,
      response: {},
    });

    const { result } = renderHook(() => useEditPost(), { wrapper: Providers });

    await result.current.editPost({
      title: 'Test Post',
      body: 'Test post body',
      slug: 'test-post',
      tags: ['tag-one'],
    });

    expect(moxios.requests.mostRecent().config.data).toMatchSnapshot();
    expect(window.location.href).contains('test-post');
  });
});
