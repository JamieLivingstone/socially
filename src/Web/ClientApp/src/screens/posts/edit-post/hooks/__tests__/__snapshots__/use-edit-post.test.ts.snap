// Vitest Snapshot v1

exports[`useEditPost > edits post and navigates to slug 1`] = `"{\\"title\\":\\"Test Post\\",\\"body\\":\\"Test post body\\",\\"slug\\":\\"test-post\\",\\"tags\\":[\\"tag-one\\"]}"`;
